internal/vehicle/carwings.go

```diff
@@ -57,52 +77,180 @@ func (v *CarWings) initSession() error {
+				err = api.ErrMustRetry
+			}
+		}
+	}
+
+	return res, err
+}
+
+// refreshResult triggers an update if not already in progress, otherwise gets result
+func (v *CarWings) refreshResult() (res carwingsVehicleStatus, err error) {
+
+	finished, err := v.session.CheckUpdate(v.refreshKey)
+	// update successful and completed
+	if err == nil && finished {
+		v.refreshKey = ""
+		var bs carwings.BatteryStatus
+		bs, err = v.session.BatteryStatus()
+		if err == nil {
+			res.VehicleStatus.SOC = float64(bs.StateOfCharge)
+			res.VehicleStatus.Range = int64(bs.CruisingRangeACOn) / 1000
+			res.VehicleStatus.ChargingStatus = bs.ChargingStatus
+			res.VehicleStatus.PlugStatus = bs.PluginState
+			res.VehicleStatus.LastRefresh = bs.Timestamp
+		}
+		var ccs carwings.ClimateStatus
+		ccs, err = v.session.ClimateControlStatus()
+		if err == nil {
+			res.VehicleStatus.ClimateStatus.active = ccs.Running
+			res.VehicleStatus.ClimateStatus.targetTemp = float64(ccs.Temperature)
+			res.VehicleStatus.ClimateStatus.outsideTemp = 0
 		}
+		// Update results here
+		return res, nil
+	}
+
+	// update still in progress, keep retrying
+	if time.Since(v.refreshTime) < carwingsRefreshTimeout {
+		return res, api.ErrMustRetry
 	}
 
-	bs, err := v.session.BatteryStatus()
-	return float64(bs.StateOfCharge), err
+	// give up
+	v.refreshKey = ""
+	if err == nil {
+		err = api.ErrTimeout
+	}
+
+	return res, err
 }
 
-// hvacAPI provides hvac-status api response
-func (v *CarWings) hvacAPI() (interface{}, error) {
+// refreshRequest requests status refresh tracked by commandId
+func (v *CarWings) refreshRequest() error {
+	// otherwise start normal workflow
 	if v.session == nil {
 		if err := v.initSession(); err != nil {
-			return 0, api.ErrNotAvailable
+			return api.ErrNotAvailable
 		}
 	}
 
-	return v.session.ClimateControlStatus()
+	var err error
+	v.refreshKey, err = v.session.UpdateStatus()
+	if err == nil {
+		v.refreshTime = time.Now()
+
+		if v.refreshKey == "" {
+			err = errors.New("refresh failed")
+		}
+	}
+
+	return err
 }
 
 // SoC implements the api.Vehicle interface
 func (v *CarWings) SoC() (float64, error) {
-	return v.chargeStateG()
+	res, err := v.statusG()
+	if res, ok := res.(carwingsVehicleStatus); err == nil && ok {
+		return float64(res.VehicleStatus.SOC), nil
+	}
+
+	return 0, err
 }
 
 var _ api.VehicleClimater = (*CarWings)(nil)
 
 // Climater implements the api.Vehicle.Climater interface
 func (v *CarWings) Climater() (active bool, outsideTemp float64, targetTemp float64, err error) {
-	res, err := v.hvacG()
-	if res, ok := res.(carwings.ClimateStatus); err == nil && ok {
-		active = res.Running
+	res, err := v.statusG()
+	if res, ok := res.(carwingsVehicleStatus); err == nil && ok {
+		active = res.VehicleStatus.ClimateStatus.active
 
-		targetTemp = float64(res.Temperature)
+		targetTemp = res.VehicleStatus.ClimateStatus.targetTemp
 
-		outsideTemp = 0 //fixed value
+		outsideTemp = res.VehicleStatus.ClimateStatus.outsideTemp
```

Comment:
its still fixed to 0- why change? would be better to set to equal targetTemp.

Reply:
carwings does not provide outside temp data. OK, I will set it to same value as targetTemp.